dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/AbstractEventService.java

```diff
@@ -1624,63 +1610,146 @@ private boolean doValidationOfMandatoryAttributes( User user )
+        // only 1 dataValue is updated and /events/{uid}/{dataElementUid} endpoint is leveraged.
+        if ( isSingleValueUpdate )
+        {
+            presentDataElements.addAll( programStageInstance.getDataValues().stream()
+                .filter( dv -> !StringUtils.isEmpty( dv.getValue() ) )
+                .map( dv -> IdentifiableObjectUtils
+                    .getIdentifierBasedOnIdScheme( dv.getDataElement(), dataElementIdScheme ) )
+                .filter( Objects::nonNull )
+                .collect( Collectors.toSet() ) );
+        }
+
+        Set<String> notPresentMandatoryDataElements = Sets
+            .difference( new HashSet<>( mandatoryDataElements.values() ), presentDataElements );
+
+        if ( !notPresentMandatoryDataElements.isEmpty() )
+        {
+            notPresentMandatoryDataElements.forEach( deIdentificator -> importSummary.getConflicts()
+                .add( new ImportConflict( deIdentificator, "value_required_but_not_provided" ) ) );
+            importSummary.incrementIgnored();
+            importSummary.setStatus( ImportStatus.ERROR );
+        }
+
+        return mandatoryDataElements;
+    }
 
-            Set<String> notPresentMandatoryDataElements = Sets.difference( mandatoryDataElements, presentDataElements );
+    private Map<String, String> validatePresenceOfMandatoryDataElementsWithAttributeIdScheme( Event event, ProgramStageInstance programStageInstance,
+        IdScheme dataElementIdScheme, ImportSummary importSummary, boolean isSingleValueUpdate )
+    {
+        Map<String, String> mandatoryDataElements = programStageInstance.getProgramStage().getProgramStageDataElements().stream()
+            .filter( ProgramStageDataElement::isCompulsory )
+            .collect( Collectors.toMap(
+                psde -> psde.getDataElement().getUid(),
+                psde -> psde.getDataElement().getName() ) );
 
-            if ( !notPresentMandatoryDataElements.isEmpty() )
+        //Using Map<UID, Attribute:UID> to be able to make a Sets.difference below
+        Map<String, String> presentDataElements = new HashMap<>();
+        event.getDataValues().stream()
+            .map( DataValue::getDataElement )
+            .forEach( de ->
             {
-                notPresentMandatoryDataElements.forEach( deUid -> importSummary.getConflicts().add( new ImportConflict( deUid, "value_required_but_not_provided" ) ) );
-                importSummary.incrementIgnored();
-                importSummary.setStatus( ImportStatus.ERROR );
-            }
+                DataElement dataElement = getDataElement( dataElementIdScheme, de );
+
+                if ( dataElement != null )
+                {
+                    presentDataElements.put( dataElement.getUid(), de );
+
+                    //If I have a correct Attribute:UID, then replace the Name byt this Attribute:UID. It will come handy later
+                    mandatoryDataElements.put( dataElement.getUid(), de );
```

Comment:
Are you sure you should put it in mandatory and present and not just present? If yes, maybe mandatory should have a different name?

Reply:
It is correct as it is. Just obtaining Attribute:UID where I can so later (in `validateDataValue()`) I can eventually show a more informative error message to the user.